plugins/single_plugins/vswipe.cpp

```diff
@@ -0,0 +1,387 @@
+#include <plugin.hpp>
+#include <output.hpp>
+#include <core.hpp>
+#include <debug.hpp>
+#include <view.hpp>
+#include <view-transform.hpp>
+#include <render-manager.hpp>
+#include <workspace-stream.hpp>
+#include <workspace-manager.hpp>
+#include <signal-definitions.hpp>
+#include <glm/glm.hpp>
+#include <glm/gtc/matrix_transform.hpp>
+#include <animation.hpp>
+
+#include <cmath>
+#include <utility>
+#include <animation.hpp>
+
+#include "vswipe-processing.hpp"
+
+class vswipe : public wf::plugin_interface_t
+{
+    private:
+        struct {
+            /* When the workspace is set to (-1, -1), means no such workspace */
+            wf::workspace_stream_t prev, curr, next;
+        } streams;
+
+        enum swipe_direction_t
+        {
+            HORIZONTAL = 0,
+            VERTICAL,
+            UNKNOWN,
+        };
+
+        struct {
+            bool swiping = false;
+            swipe_direction_t direction;
+
+            wf_pointf initial_deltas;
+
+            double delta_sum = 0.0;
+            double gap = 0.0;
+
+            double delta_prev = 0.0;
+            double delta_last = 0.0;
+
+            int vx = 0;
+            int vy = 0;
+            int vw = 0;
+            int vh = 0;
+        } state;
+
+        wf::render_hook_t renderer;
+
+        wf_duration duration;
+        wf_transition transition;
+
+        wf_option animation_duration;
+        wf_option background_color;
+        wf_option enable_horizontal;
+        wf_option enable_vertical;
+        wf_option ignore_cancel;
```

Comment:
Can you explain when cancel happens?

Reply:
grep for the usage of this :) `(!ev->cancelled || ignore_cancel->as_int())` — `cancelled` is a property of the `end` event. libinput has some weird heuristics for whether the gesture was "cancelled", in my experience it doesn't make sense for this plugin, it was usually true when I don't want it.

I guess we can just ignore it always and not have the setting.

Reply:
in my tests it also didn't seem necessary to change it. So maybe we should just ignore it always.